### ISSUE TITLE
Unexpected behavior with LuaLock

### ISSUE THREAD

**Issue by mikeyk** [alisaifee/coredis#113]
The docs say under LuaLock:

> blocking – indicates whether calling acquire() should block until the lock has been acquired or to fail immediately, causing acquire() to return False and the lock not being acquired. Defaults to True.

However, if `blocking=False`, and two processes attempt to grab the lock, they both run the code inside (there's nothing in the LuaLock that prevents concurrent execution). Though we return `False` from `acquire()`, `__aenter__` executes the wrapped code regardless. As a brief PoC, this code run in parallel in two procs will repro:

```
async with LuaLock(name="demo_lock", blocking=False, timeout=15, client=redis):
    print("I'm inside!")
    await asyncio.sleep(10)
```

Also, the second one (which didn't grab the lock) will throw a `LockError: Cannot release an unlocked lock` because it tries to call release, but doesn't actually have the lock token.

I'm happy to take a crack at a PR here but wanted to align on what it should do:
1. Raise some sort of `LockNotAcquired` error on acquisition, since that's the only way to bail out of `__aenter__` without executing the code inside
2. Yield a boolean `acquired` (instead of yielding the `LuaLock`) indicating whether we acquired the lock or not, that can then be checked inside the wrapped code
3. Add a `succeeded` property to the existing yielded `LuaLock` object that can be checked inside the wrapped code.

For 2 or 3 we'd also need to only call `release` if we actually hold the key.

I think 3 is likely best, as existing code might rely on getting the `LuaLock` as the yielded value of the async context manager. Though 1 is the safest and probably will expose some existing bugs out in the wild.

**Comment by alisaifee** [comment]
> **Note**
I'd originally removed the lock implementation when adopting the codebase from [aredis](https://github.com/NoneGG/aredis) and only back ported a stripped down variant due to a request in #71.

I think you're right that the safest and smallest change can be 1 (basically add an `if` [here](https://github.com/alisaifee/coredis/blob/master/coredis/recipes/locks/lua_lock.py#L131))

**Comment by alisaifee** [comment]
Since the lock implementation is in the `recipes` namespace I think (hope?) that making the change to throw a `LockError` in `__aenter__` as the correct behaviour is fine instead of worrying about backward compatibility.

**Comment by mikeyk** [comment]
Great! I'll PR that up.

**Comment by alisaifee** [comment]
Nice, I'm lining up a `4.7.0` release so I'll hold off till we get your PR merged!

**Comment by alisaifee** [comment]
The fix will be available in 4.7.0.
